compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -1621,6 +1621,36 @@ impl<'a, 'tcx> InferCtxtPrivExt<'a, 'tcx> for InferCtxt<'a, 'tcx> {
                 "type mismatch resolving `{}`",
                 predicate
             );
+            let mut v = HirFinder { needle: obligation.cause.span, found: None };
+            let body = self.tcx.hir().get(obligation.cause.body_id);
+            if let hir::Node::Expr(expr) = body {
+                v.visit_expr(expr);
+            }
+            if let Some(expr) = v.found {
+                // FIXME(estebank): look for method calls that can be applied on the previous
+                // element of the chain, that would make the next call to typeck.
+                let mut receiver = expr;
+                if let (Some(typeck_results), hir::ExprKind::MethodCall(..)) =
+                    (self.in_progress_typeck_results, &receiver.kind)
+                {
+                    let typeck_results = typeck_results.borrow();
+                    while let hir::ExprKind::MethodCall(method_segment, args, _) = receiver.kind {
+                        if let Some(ty) = typeck_results.expr_ty_adjusted_opt(receiver) {
```

Comment:
I wonder if resolving `ty` here would give us the `Item` type instead of `_`, which *would* be useful then.